python/python_fastapi_server/routers/caching_middleware.py

```diff
@@ -0,0 +1,143 @@
+from io import BytesIO
+import os
+from urllib.parse import urlsplit
+
+from starlette.middleware.base import BaseHTTPMiddleware
+from starlette.responses import Response
+from fastapi import BackgroundTasks
+
+import calendar
+from datetime import datetime
+import redis.asyncio as redis  # This can also be used to connect to a Redis cluster
+
+# from redis.asyncio.cluster import RedisCluster as Redis  # Cluster client, for testing
+
+import json
+import brotli
+
+ADAGUC_REDIS = os.environ.get("ADAGUC_REDIS")
+
+
+async def get_cached_response(redis_pool, request):
+    key = generate_key(request)
+    redis_client = redis.Redis(connection_pool=redis_pool)
+    cached = await redis_client.get(key)
+    await redis_client.aclose()
+    if not cached:
+        return None, None, None
+
+    entrytime = int(cached[:10].decode("utf-8"))
+    currenttime = calendar.timegm(datetime.utcnow().utctimetuple())
+    age = currenttime - entrytime
+
+    headers_len = int(cached[10:16].decode("utf-8"))
+    headers = json.loads(cached[16 : 16 + headers_len].decode("utf-8"))
+
+    data = brotli.decompress(cached[16 + headers_len :])
+    return age, headers, data
+
+
+skip_headers = ["x-process-time", "age"]
```

Comment:
duplicate code?

Reply:
Renamed here too.
I don't see this asduplicate code, because the caching_middleware and runAdaguc use the cache slightly differently

Reply:
done